src/ipc/ipc4/logging.c

```diff
@@ -84,22 +84,30 @@ static void mtrace_log_hook(size_t written, size_t space_left)
 		ipc_send_buffer_status_notify();
 		mtrace_notify_last_sent = k_uptime_get();
 		mtrace_bytes_pending = 0;
+	} else {
+		k_work_schedule_for_queue(&ipc_get()->ipc_send_wq, &log_work,
+					  K_MSEC(mtrace_aging_timer - delta));
 	}
 }
 
-static enum task_state mtrace_task_run(void *data)
+static void mtrace_log_hook(size_t written, size_t space_left)
 {
-	if (k_uptime_get() - mtrace_notify_last_sent >= mtrace_aging_timer &&
-	    mtrace_bytes_pending)
-		mtrace_log_hook(0, 0);
+	k_mutex_lock(&log_mutex, K_FOREVER);
+
+	mtrace_log_hook_unlocked(written, space_left);
 
-	/* task will be re-run based on mtrace_task_deadline() */
-	return SOF_TASK_STATE_RESCHEDULE;
+	k_mutex_unlock(&log_mutex);
 }
 
-static uint64_t mtrace_task_deadline(void *data)
+static void log_work_handler(struct k_work *work)
 {
-	return k_uptime_ticks() + k_ms_to_ticks_ceil64(mtrace_aging_timer);
+	k_mutex_lock(&log_mutex, K_FOREVER);
+
+	if (k_uptime_get() - mtrace_notify_last_sent >= mtrace_aging_timer &&
```

Comment:
Why you need to check this? Shouldn't Zephyr's Delayable Work call this handler no sooner than a timeout?

Reply:
.. and there is yet another check in mtrace_log_hook_unlocked(). This check look quite redundant to me too.

Reply:
I think both are needed. This one is needed because it could happen that after the work was queued another log was issued and then the space was tight so an IPC got sent out and the time updated, so now the time isn't larger than the threshold. And the other one is needed for normal hook operation - if it isn't called from the delayed work, but enough time has passed since the last notification, we'll notify now.